src/compiler/parser.ts

```diff
@@ -1021,6 +1052,121 @@ namespace ts {
             return node;
         }
 
+        function reparseTopLevelAwait(sourceFile: SourceFile) {
+            return visitEachChild(sourceFile, visitor, reparseContext);
+
+            function visitor(node: Node): VisitResult<Node> {
+                if (!(node.transformFlags & TransformFlags.ContainsPossibleTopLevelAwait)) {
+                    return node;
+                }
+                switch (node.kind) {
+                    case SyntaxKind.Decorator: return reparseDecorator(node as Decorator);
+                    case SyntaxKind.ComputedPropertyName: return reparseComputedPropertyName(node as ComputedPropertyName);
+                    case SyntaxKind.ExpressionWithTypeArguments: return reparseExpressionWithTypeArguments(node as ExpressionWithTypeArguments);
+                    case SyntaxKind.ExpressionStatement: return reparseExpressionStatement(node as ExpressionStatement);
+                    case SyntaxKind.IfStatement: return reparseIfStatement(node as IfStatement);
+                    case SyntaxKind.SwitchStatement: return reparseSwitchStatement(node as SwitchStatement);
+                    case SyntaxKind.WithStatement: return reparseWithStatement(node as WithStatement);
+                    case SyntaxKind.DoStatement: return reparseDoStatement(node as DoStatement);
+                    case SyntaxKind.WhileStatement: return reparseWhileStatement(node as WhileStatement);
+                    case SyntaxKind.ForStatement: return reparseForStatement(node as ForStatement);
+                    case SyntaxKind.ForInStatement: return reparseForInStatement(node as ForInStatement);
+                    case SyntaxKind.ForOfStatement: return reparseForOfStatement(node as ForOfStatement);
+                    case SyntaxKind.ReturnStatement: return reparseReturnStatement(node as ReturnStatement);
+                    case SyntaxKind.ThrowStatement: return reparseThrowStatement(node as ThrowStatement);
+                    case SyntaxKind.ExportAssignment: return reparseExportAssignment(node as ExportAssignment);
+                    case SyntaxKind.VariableDeclaration: return reparseVariableDeclaration(node as VariableDeclaration);
+                    case SyntaxKind.BindingElement: return reparseBindingElement(node as BindingElement);
+                    default: return visitEachChild(node, visitor, reparseContext);
+                }
+            }
+
+            function reparse<T extends Node>(node: T, parse: () => T): T | Expression;
+            function reparse<T extends Node>(node: T | undefined, parse: () => T): T | Expression | undefined;
+            function reparse<T extends Node>(node: T | undefined, parse: () => T) {
+                if (node && node.transformFlags & TransformFlags.ContainsPossibleTopLevelAwait) {
+                    if (isExpression(node)) {
+                        return speculationHelper(() => {
+                            scanner.setTextPos(node.pos);

```

Comment:
Would using `initializeState` and passing in a syntax cursor here allow us to reuse the initially parsed nodes where possible, even in the case of a partial reparse? (Reparse due to context invalidation and incremental reparsing I would think seem similar)

Reply:
Possibly. I'm on the fence as to whether I wanted to set `original` pointers for these nodes. I need to add some incremental parse test cases.

Reply:
I considered using incremental parse, but it doesn't apply at the expression level, only at certain list context levels (see https://github.com/microsoft/TypeScript/pull/39084#discussion_r441770690). I might have been able to leverage it by reparsing the statement itself, but some of the necessary incremental parse machinery isn't currently accessible from within `Parser`, and would be fairly complex to wire up properly.